### PR TITLE
[Marvell-arm64]: Fix sonic-installer

### LINKED ISSUE #18681: [Marvell-arm64] Sonic install gives illegal character error
#### Description
Sonic install give error of illegal characters. After which sonic install did not execute all the steps.

Error: illegal character '=' in variable name "rootfstype=ext4"
Error: illegal character '=' in variable name "rootfstype=ext4"

#### Steps to reproduce the issue:
1. sonic-installer install <sonic-marvell-arm64.bin> -y

#### Describe the results you received:
sonic-installer install  
Installing image SONiC-OS-202311.0-dirty-20240131.102205 and setting it as default...
Command: bash /tmp/sonic_image
 
Error: illegal character '=' in variable name "rootfstype=ext4"
Error: illegal character '=' in variable name "rootfstype=ext4"
Verifying image checksum ... OK.
Preparing image archive ... OK.
Installing SONiC in SONiC
ONIE Installer: platform: arm64-marvell-arm64-r0
onie_platform: arm64-marvell_rd98DX35xx-r0
Preparing for installation ...
Intalling SONiC from sonic on Platform arm64-marvell_rd98DX35xx-r0
/dev/mmcblk0
Selected mmc /dev/mmcblk0
Removing old SONiC installation /host/image-HEAD.0-dirty-20240414.072620
Installing SONiC to /host/image-202311.0-dirty-20240131.102205
Archive:  fs.zip
   creating: /host/image-202311.0-dirty-20240131.102205/boot/
  inflating: /host/image-202311.0-dirty-20240131.102205/boot/System.map-5.10.0-23-2-arm64 
  inflating: /host/image-202311.0-dirty-20240131.102205/boot/config-5.10.0-23-2-arm64 
  inflating: /host/image-202311.0-dirty-20240131.102205/boot/initrd.img-5.10.0-23-2-arm64 
  inflating: /host/image-202311.0-dirty-20240131.102205/boot/vmlinuz-5.10.0-23-2-arm64 
  inflating: /host/image-202311.0-dirty-20240131.102205/boot/cn9130-db-comexpress.dtb 
  inflating: /host/image-202311.0-dirty-20240131.102205/boot/cn9130-db-comexpress.dts 
  inflating: /host/image-202311.0-dirty-20240131.102205/boot/cn9131-db-comexpress.dtb 
  inflating: /host/image-202311.0-dirty-20240131.102205/boot/cn9131-db-comexpress.dts 
  inflating: /host/image-202311.0-dirty-20240131.102205/boot/sonic_fit.its 
  inflating: /host/image-202311.0-dirty-20240131.102205/boot/sonic_arm64.fit 
 extracting: /host/image-202311.0-dirty-20240131.102205/fs.squashfs 
ONIE_IMAGE_PART_SIZE=32768
EXTRA_CMDLINE_LINUX=
Sync up cache ...
Setting up U-Boot environment...
Using pre-configured uboot env

#### Output of `show version`:

```
(paste your output here)
```
#### Output of `show techsupport`:

```
(paste your output here or download and attach the file here )
```

### PR DESCRIPTION
**Pull request by pavannaregundi**
fixes #18681

#### Why I did it
Fix sonic installer issues for marvell-arm64 platform. Following fixes are part of change,
- Fix illegal character error
- Set second image based on current image
- Fix boot_once uboot environment variable
- Fix print boot menu

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it
Changed the platform.conf scripts which sets uboot environment variables.
 
#### How to verify it
Tested following cases to verify the issue,
 - sonic-installer install
 - sonic-installer set-next-boot
 - sonic-installer set-default

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

**Comment by pavannaregundi** [comment]
@Blueve @Pavan-Nokia Please review.

File: platform/marvell/platform_arm64.conf [hunk 1]
```diff
@@ -231,14 +242,14 @@ prepare_boot_menu() {
...
-    fw_setenv ${FW_ARG} sonic_bootargs $BOOTARGS > /dev/null
-    fw_setenv ${FW_ARG} sonic_bootcmd $SONIC_BOOT_CMD > /dev/null
-    fw_setenv ${FW_ARG} sonic_image_2 $SONIC_BOOT_CMD_OLD > /dev/null
+    BOOTARGS='setenv bootargs root='$demo_dev' rw rootwait panic=1 console=ttyS0,${baudrate} ${othbootargs} ${mtdparts} ${linuxargs}'
+    fw_setenv ${FW_ARG} sonic_bootargs "$BOOTARGS" > /dev/null
+    fw_setenv ${FW_ARG} sonic_image_2 "$SONIC_BOOT_CMD_OLD" > /dev/null
     fw_setenv ${FW_ARG} sonic_image_1 "$SONIC_BOOT_CMD" > /dev/null
     fw_setenv ${FW_ARG} boot_next  'run sonic_image_1'> /dev/null
-    fw_setenv ${FW_ARG} bootcmd 'run print_menu; usb start; test -n "$boot_once" && run boot_once; run boot_next' > /dev/null
+    fw_setenv ${FW_ARG} bootcmd 'run print_menu; usb start; test -n "$boot_once" && setenv boot_once "" && saveenv && run boot_once; run boot_next' > /dev/null
```
Comment (Pavan-Nokia) [inline platform/marvell/platform_arm64.conf:250]: This change looks incorrect. the boot_once variable should not be cleared before running it. the boot once will never take effect.

The boot_once variable must be cleared in SONIC_BOOT_CMD and SONIC_BOOT_CMD_OLD

File: platform/marvell/platform_arm64.conf [hunk 2]
```diff
@@ -212,12 +223,12 @@ prepare_boot_menu() {
...
     BOOT3='echo " > Boot3: ONIE - run onie_boot";echo;'
     BORDER='echo "---------------------------------------------------";echo;'
-    fw_setenv ${FW_ARG} print_menu $BORDER $BOOT1 $BOOT2 $BOOT3 $BORDER > /dev/null
+    fw_setenv ${FW_ARG} print_menu "$BORDER $BOOT1 $BOOT2 $BOOT3 $BORDER" > /dev/null
 
     fw_setenv ${FW_ARG} linuxargs "net.ifnames=0 loopfstype=squashfs loop=$image_dir/$FILESYSTEM_SQUASHFS systemd.unified_cgroup_hierarchy=0 varlog_size=$VAR_LOG ${ONIE_PLATFORM_EXTRA_CMDLINE_LINUX}" > /dev/null
     fw_setenv ${FW_ARG} linuxargs_old "net.ifnames=0 loopfstype=squashfs loop=$image_dir_old/$FILESYSTEM_SQUASHFS systemd.unified_cgroup_hierarchy=0 varlog_size=$VAR_LOG ${ONIE_PLATFORM_EXTRA_CMDLINE_LINUX}" > /dev/null
-    sonic_bootargs_old='setenv bootargs root='$demo_dev' rw rootwait rootfstype=ext4 panic=1 console=ttyS0,${baudrate} ${othbootargs} ${mtdparts} ${linuxargs_old}'
-    fw_setenv ${FW_ARG} sonic_bootargs_old $sonic_bootargs_old > /dev/null || true
+    sonic_bootargs_old='setenv bootargs root='$demo_dev' rw rootwait panic=1 console=ttyS0,${baudrate} ${othbootargs} ${mtdparts} ${linuxargs_old}'
```
Comment (Pavan-Nokia) [inline platform/marvell/platform_arm64.conf:230]: any reason for removing ext4 file type

File: platform/marvell/platform_arm64.conf [hunk 3]
```diff
@@ -187,12 +187,23 @@ prepare_boot_menu() {
...
+        CURR_SONIC_IMAGE="$(sonic-installer list | grep "Current: " | cut -f2 -d' ')"
+        FIRST_SONIC_IMAGE="$(fw_printenv sonic_version_1 | cut -f2 -d'=')"
+        if [ "$CURR_SONIC_IMAGE" = "$FIRST_SONIC_IMAGE" ]; then
+            image_dir_old=$(fw_printenv -n image_dir || true)
+            image_name_old=$(fw_printenv -n image_name || true)
+            initrd_name_old=$(fw_printenv -n initrd_name || true)
+            fdt_name_old=$(fw_printenv -n fdt_name || true)
+            fit_name_old=$(fw_printenv -n fit_name || true)
+            sonic_version_2=$(fw_printenv -n sonic_version_1 || true)
+        else
```
Comment (Pavan-Nokia) [inline platform/marvell/platform_arm64.conf:199]: This else seems unnecessary as we are writing the same values back

File: platform/marvell/platform_arm64.conf [hunk 4]
```diff
@@ -231,14 +242,14 @@ prepare_boot_menu() {
     fw_setenv ${FW_ARG} sonic_boot_load "$MMC_LOAD" > /dev/null
     SONIC_BOOT_CMD='run sonic_bootargs; run sonic_boot_load; bootm $fit_addr${fit_conf_name}'
     SONIC_BOOT_CMD_OLD='run sonic_bootargs_old; run sonic_boot_load_old; bootm $fit_addr${fit_conf_name}'
-    BOOTARGS='setenv bootargs root='$demo_dev' rw rootwait rootfstype=ext4 panic=1 console=ttyS0,${baudrate} ${othbootargs} ${mtdparts} ${linuxargs}'
-    fw_setenv ${FW_ARG} sonic_bootargs $BOOTARGS > /dev/null
-    fw_setenv ${FW_ARG} sonic_bootcmd $SONIC_BOOT_CMD > /dev/null
-    fw_setenv ${FW_ARG} sonic_image_2 $SONIC_BOOT_CMD_OLD > /dev/null
+    BOOTARGS='setenv bootargs root='$demo_dev' rw rootwait panic=1 console=ttyS0,${baudrate} ${othbootargs} ${mtdparts} ${linuxargs}'
```
Comment (Pavan-Nokia) [inline platform/marvell/platform_arm64.conf:245]: same comment as above

File: platform/marvell/platform_arm64.conf (hunk 3 above)
Comment (pavannaregundi) [reply platform/marvell/platform_arm64.conf:199]: Will need to move the fw_setenv for old variables inside the if check to remove else. Will do it.

File: platform/marvell/platform_arm64.conf (hunk 2 above)
Comment (pavannaregundi) [reply platform/marvell/platform_arm64.conf:230]: Filesystem type is by default ext4. Explicit filesystem type is not required.

File: platform/marvell/platform_arm64.conf (hunk 4 above)
Comment (pavannaregundi) [reply platform/marvell/platform_arm64.conf:245]: Filesystem type is by default ext4. Explicit filesystem type is not required.

File: platform/marvell/platform_arm64.conf (hunk 1 above)
Comment (pavannaregundi) [reply platform/marvell/platform_arm64.conf:250]: Ya. We should be storing the boot_once is another variable and use it during run. Will do that changes.

File: platform/marvell/platform_arm64.conf [hunk 5]
```diff
@@ -231,14 +236,14 @@ prepare_boot_menu() {
...
+    fw_setenv ${FW_ARG} sonic_bootargs "$BOOTARGS" > /dev/null
+    fw_setenv ${FW_ARG} sonic_image_2 "$SONIC_BOOT_CMD_OLD" > /dev/null
     fw_setenv ${FW_ARG} sonic_image_1 "$SONIC_BOOT_CMD" > /dev/null
     fw_setenv ${FW_ARG} boot_next  'run sonic_image_1'> /dev/null
-    fw_setenv ${FW_ARG} bootcmd 'run print_menu; usb start; test -n "$boot_once" && run boot_once; run boot_next' > /dev/null
+    fw_setenv ${FW_ARG} bootcmd 'run print_menu; usb start; test -n "$boot_once" && setenv do_boot_once "$boot_once" && setenv boot_once "" && saveenv && run do_boot_once; run boot_next' > /dev/null
```
Comment (Pavan-Nokia) [inline platform/marvell/platform_arm64.conf:244]: Can we please use the same approach as in platform_armhf.conf for this.
that approach has been well tested

Comment (pavannaregundi) [reply platform/marvell/platform_arm64.conf:245]: This approach is also taken from other vendors and is tested to work.

**Review by Pavan-Nokia — APPROVED** [review]

**Comment by pavannaregundi** [comment]
@yxieca @Blueve All comments are addressed. Can we merge this?

**Pull request merged**